src/ray/core_worker/core_worker.cc

```diff
@@ -642,31 +673,34 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
     return_ids.pop_back();
     task_type = TaskType::ACTOR_TASK;
   }
-  bool direct_call = worker_context_.CurrentActorUseDirectCall();
-  status = task_execution_callback_(
-      task_type, func, task_spec.GetRequiredResources().GetResourceMap(), args,
-      arg_reference_ids, return_ids, direct_call, results);
 
-  SetCurrentTaskId(TaskID::Nil());
-  worker_context_.ResetCurrentTask(task_spec);
+  std::vector<std::shared_ptr<RayObject>> return_objects;
+  status = task_execution_callback_(task_type, func,
+                                    task_spec.GetRequiredResources().GetResourceMap(),
+                                    args, arg_reference_ids, return_ids, &return_objects);
 
-  // TODO(edoakes): this is only used by java.
-  if (results->size() != 0 && !direct_call) {
-    for (size_t i = 0; i < results->size(); i++) {
-      ObjectID id = ObjectID::ForTaskReturn(
-          task_spec.TaskId(), /*index=*/i + 1,
-          /*transport_type=*/static_cast<int>(TaskTransportType::RAYLET));
-      if (!Put(*results->at(i), id).ok()) {
-        // NOTE(hchen): `PlasmaObjectExists` error is already ignored inside
-        // Put`, we treat other error types as fatal here.
-        RAY_LOG(FATAL) << "Task " << task_spec.TaskId() << " failed to put object " << id
-                       << " in store: " << status.message();
-      } else {
-        RAY_LOG(DEBUG) << "Task " << task_spec.TaskId() << " put object " << id
-                       << " in store.";
+  for (size_t i = 0; i < return_objects.size(); i++) {
+    // The object is nullptr if it already existed in the object store.
+    if (!return_objects[i]) {
+      continue;
+    }
+    if (return_objects[i]->GetData()->IsPlasmaBuffer()) {
+      if (!Seal(return_ids[i]).ok()) {
```

Comment:
`Seal()` should be moved into the execution callback itself, in the same function where `Create()` is called.

Reply:
I think that's weird since it exposes the type of the buffer to the language runtime. This way, it is handled transparently.

Reply:
Makes sense. I was just wondering if there're a way to put the `Create()` and `Seal()` together, right now there's an assumption that user must call `AllocateReturnObjects` to allocate the buffers otherwise things will break. But that might not be possible since we want to avoid the an extra copy of the buffer for plasma.